### DISCUSSION TITLE
Marlin firmware source code is incomplete

### DISCUSSION THREAD

**Discussion by that-ambuj** [1coderookie/Kobra2NeoInsights#5] [General]
As mentioned in this section: https://1coderookie.github.io/Kobra2NeoInsights/firmware/fw_marlin/ One needs to follow the guide on reddit with Marlin source code from Anycubic at hand. 

However, halfway through following the guide I realised that the file `anycubic.uvprojx` is missing from the source code and I realised that the guide only works for Kobra Neo and the code for Kobra 2 Neo cannot be compiled or has to be compiled in some other way. 

I want to increase the serial baudrate and the buffer size for my printer which is probably too low by default (baudrate: 115200) because I'm facing stuttering issue while running octoprint on a fairly powerful phone. 

If anyone knows how to compile the Kobra 2 Neo firmware, please help me out here. 

@1coderookie Please double check that my assumption is right and update the infosite accordingly. It's been really useful to me and I don't want others to be misguided by it, Thanks for the good work.

**Comment by 1coderookie** [comment]
Hi, 
thanks for reaching out!  
This is interesting, but I'm afraid I can't help. I can't check on it as I don't have the K2 Neo and I don't have the program for the Marlin stuff as well. You could open an issue at AC's GH repo, but I doubt they'll add it tbh (I personally switched to Klipper and I'd suggest to do the same in your case, if you really think it's an issue with what you mentioned), but I'll add a note and link to this discussion.  
But: I don't really see where it should be caused by the baudrate or the buffer size of the USB connection - even when using OctoPrint, the whole file gets send to the printer first, then it reads the data from there. Unlike Klipper, where it sends the commands one after another.  

Does it happen with every print? Or only with certain complex files or models or so?   
And what *exactly* happens?

**Reply by that-ambuj** [reply]
In my case, the whole file is not sent to the printer's sd card unless I explicitly upload the files to the SD Card. I'm using Orcaslicer and by default it uploads the file to the octoprint server/device and that device sends those commands one by one via the serial connection, similar to how you mention how klipper works.

It usually stutters when printing a lot of curves/circles, except the gyroid infill. Try the retraction test with two cylinder found in the Orcaslicer. I've tried using a different slicer. I've enabled Arc fitting so it does send G2/G3 gcode for the curves so that doesn't seem to be enough, I've tried the arc welder plugin and there's no difference there. 

Exactly what happens is that the printhead freezes for like 0.5 seconds and does not retract. I've noticed it happens on a layer change, I might be wrong in that assumption. Since it's not retracting, I'm pretty sure it's not intentional for timelapse or something else. I have disabled timelapse for that reason, it was happening even when I initially disabled the webcam because I didn't have a tripod at the time and that makes me confident that the phone is not lagging behind because of webcam or something else. I also tried safemode on octoprint and that didn't help. The resend ratio is always 0 so that doesn't indicate any issues.

I'm considering installing Klipper because that might just fix the issue. Can I know what baudrate is set in your klipper firmware? 

Also, completely unrelated, the printer's z-offset is very different every single print it's about 0.2 - 0.4mm off everytime and the ABL is always wrong, do you have any idea about that? I'll created a seperate discussion for that if it's not an issue.

**Reply by 1coderookie** [reply]
Ah I see.. Hmm, and does it happen as well when you upload to and print from the SD card? 
Because I'd be curious if it really is a problem with e.g. the baudrate or buffer, or if it's because of the complexity of the model. 
Did you try printing slower as well? 

Klipper baudrate: depends on what you set. I used both 250000 as well as 112500 w/o any problems. Iirc, I have most of my machines set to 115200 tho..   

ABL: yes, please open another discussion for that then please..

**Reply by that-ambuj** [reply]
I did not try printing slower but the slicer makes object print slower by default because of the amount of cooling needed and the issue was still there. I have also tried switching the usb cable to a new one.

**Reply by 1coderookie** [reply]
I'd try printing the exact same file from the card as well..

**Reply by that-ambuj** [reply]
Thanks, I didn't think of that, let me quickly do that and report back.

**Reply by that-ambuj** [reply]
I just started printing with the SD Card and there are 0 hiccups or stutters so far. 

Update: just printed and observed the same file and it's printing perfectly fine now.

**Reply by that-ambuj** [reply]
@1coderookie I think you're not facing the issue with the baudrate because you previously mentioned you print from the SD Card and I'm assuming you do that for the other printers, you probably won't have experienced the same thing.

**Reply by 1coderookie** [reply]
No, I'm using Klipper at all of my machines, I'm not printing from SD at all. I said I *would* do so, to check if the issue is the same since you didn't answer my questions if you already did that.

**Reply by that-ambuj** [reply]
> I just started printing with the SD Card and there are 0 hiccups or stutters so far.
> 
> Update: just printed and observed the same file and it's printing perfectly fine now.

@1coderookie I answered the question in the previous reply. There's no issue printing the same file from sd card but the issue is there on octoprint.

**Reply by 1coderookie** [reply]
Yes, I saw that, but before that you didn't and that's why I wrote "I'd still check..".. ;)
Anyway, if that works now, then at least it's clear that it's really related to your OctoPrint solution somehow.

**Comment by 1coderookie** [comment]
I just added a warning box: https://1coderookie.github.io/Kobra2NeoInsights/firmware/fw_marlin/  

@jojos38 Do you maybe know anything about this problem with that missing file and what to do about it..?  
Seems that someone else (or is it you @that-ambuj ?) came across the same issue at [reddit](https://www.reddit.com/r/anycubic/comments/1gcsrzj/platformioini/?rdt=51129), I must have missed that one.. :(

**Reply by that-ambuj** [reply]
It is someone else on reddit and seeing the post increased my suspicions. 

There is also an open issue in the Marlin repo https://github.com/MarlinFirmware/Marlin/issues/26682, in which one of the users compiled the firmware without a printer but they deleted the branch (for security reasons, I assume?).

**Reply by 1coderookie** [reply]
Ah ok, so you didn't reach out to AC yet? Neither to the support nor by opening an issue at their GH repo?

**Reply by that-ambuj** [reply]
Not yet, because I don't think compiling firmware is part their support coverage? Besides, the point of this discussion was to at the very least to correct the infosite. 

I'll soon create an issue on their GitHub and link it here. 

In my opinion, their repo does not have a single sentence of documentation on how to build the firmware which makes the attempt to ask for help even more futile.

**Reply by 1coderookie** [reply]
Got it, thanks again for reaching out! 
True, but if really files are missing which would *necessary* to build their sources, then they'd need/have to add those I reckon? Dunno how it's about that in terms of GPL tho.. Sure, hopes aren't high that they'll fix it, but we never know..

**Reply by that-ambuj** [reply]
There's only one way to find out and I might as well just do it.

**Reply by 1coderookie** [reply]
If you have a reddit account, maybe make a post there as well, like "hey, any news, I have the same problem". Then I'll make one as well where I mention some ppl who work for AC, maybe that will help.

**Reply by that-ambuj** [reply]
I just commented on the post, let's see if they found something. I think most likely not.

**Reply by that-ambuj** [reply]
@1coderookie Hey! I have some news from the reddit post. The guy replied and he said that he has the same issue with stuttering in octoprint, I don't think that's a coincidence. 

Here's the link: https://www.reddit.com/r/anycubic/s/xaa63frRs2

**Comment by GNN3** [comment]
.I still managed to compile Marlin on Anicubic Cobra 2 Neo.
There is [one missing file](https://github.com/ANYCUBIC-3D/framework-arduinoststm32_kobra2_neo/issues/1) in the drivers for this printer on github.

**Reply by GNN3** [reply]
[Here is a small assembly guide](https://4pda.to/forum/index.php?showtopic=1085070&view=findpost&p=133323862)

**Reply by that-ambuj** [reply]
@1coderookie What do you think about this? I'm going to try this.

**Reply by that-ambuj** [reply]
@GNN3 Thanks a lot for this. I actually worked!

The missing platformio.ini is now added to the repo but the files for the board are missing from platformio extension for vscode itself, I think there should be a way to add that through a PR.

@1coderookie I'm still finding ways to modify the firmware in order to fix the stuttering issue, I'll update you when I fix it.

**Reply by 1coderookie** [reply]
@GNN3 @that-ambuj  Sorry for being so late guys, but thanks for the info!

**Reply by that-ambuj** [reply]
@1coderookie So I think you should mark @GNN3's reply as the answer and probably update the infosite to include the guide in a translated version.

**Comment by that-ambuj** [comment]
@1coderookie I am now able to compile the firmware thanks to @GNN3 and I fixed the issue with stuttering.

In order to fix the stuttering, I had to enable MEATPACK in firmware and add the respective plugin in Octoprint. I also increased the buffer sizes for good measure and enabled some features I wanted. However, just to warn others, increasing the baudrate made things worse.

All the fixes are available [here](https://github.com/that-ambuj/Kobra2_Neo/tree/octoprint-stutter-fix) for people to compile themselves.

**Reply by GNN3** [reply]
> I fixed the issue with stuttering.

Happy New Year!
I use the Meatpack and the firmware you modified, but still in some cases (especially at high print speeds) the printer stutters slightly, including when working with an SD card. It seems to me that this is due to the default function of resuming printing after the power disappears (M413).
The printer is constantly trying to write its status to a very slow SD card, which simply cannot but cause problems with the already slow hardware.
I'm currently testing a printer with a g-code in a slicer:
M413 S0 
M500
And it seems the stuttering has disappeared.

**Reply by kmars42** [reply]
I'm late to the game on this one but I'm trying to get caught up.  I managed to get the firmware to compile thanks to the instructions provided by @GNN3 and install correctly but I'm having Z position issues (they existed before the new version of Marlin) so I can't try and print yet.  Everything seems to be working as expected and I can move X, Y, Z and the extruder as they should.  I have a question though, and I wonder if it may have something to do with the stuttering others are talking about.  In the source provided by AnyCubic they have the stepper driver types set to A4988 but everything I've found online says that these motherboards use TMC2209 drivers.  I would think that would be an important difference.  Am I missing something?

**Reply by 1coderookie** [reply]
It seems that AC just set the wrong stepper driver there, at least based on what we know about the type of stepper drivers being used.

**Reply by kmars42** [reply]
Turns out that the chips on the board are actually GC6609's.  I found a discussion about them on github where someone says that they appear to be a lower cost clone of the TMC2209 but that they're probably more like the TMC2208.  Here's a link:

https://github.com/1coderookie/KobraGoNeoInsights/discussions/13

So my main concern is how much difference does it make in Marlin?  I built it with the drivers set to A4988 as AnyCubic had it and I also built it with them set as TMC2209's and I can't tell any difference.

**Reply by 1coderookie** [reply]
How did it turn out? Did you pull off the heatsinks from the stepper drivers of your mobo?  
If not, then it still could be either TMC2208/TMC2209/GC6609. No matter which one is on your mobo (and it's pretty likely that they changed during mobo-revs like at the Kobra Neo as well, where the early ones had TMC2208 and later ones had GC6609 for example), I'd say it's definitely not A4988.  
I don't know how or where it would make a difference in Marlin, probably there isn't any (to notice) until you (try to) really use the advanced TMC features, as it apparently works with A4988 instead of TMCs being set at this point.  

Out of curiousity: what are you trying to achieve with modifying the stock fw and why don't you use Klipper instead?

**Reply by kmars42** [reply]
I did pull a heatsink off because I replaced the motherboard.  I had an issue that I couldn't track down and thought it might be the motherboard and, although they don't list them on their site, they do sell them on eBay for $19 so I took a chance.  It didn't help the problem but I pulled a heatsink off one of the chips on the original board and it was a GC6609 and both motherboards are V4.0.1 so I expect they use the same chips.  The only reason I was trying to build the firmware is to enable HOST_ACTION_COMMANDS so that it would work better with OctoPrint.  I've built it with the chips set as A4988 and as TMC2209 and I don't notice any difference but I'm not yet knowledgeable enough to do anything beyond the basics at this point.  I was more concerned with causing damage rather than having access to better features.  Interestingly though, the new motherboard has a newer version of the firmware than the old one and the old one had the newest version that was available on their site.  It still doesn't support HOST_ACTION_COMMANDS so I went to their site to see if I can find out what they changed and they currently don't have ANY firmware or information available and there's a place-holder saying to wait for the new version.  The version that was formerly on their site is V1.5.6.3 and it was still available there a couple of weeks ago.  The version on the new motherboard is V1.5.6.4.

**Reply by 1coderookie** [reply]
So we have proof about the GC6609 then!?! That's great!  

HOST_ACTION_COMMANDS - something rings a bell in the back of my head, but I don't use any Marlin anymore, only Klipper at all of my printers. 
Maybe give Klipper a try as well, I really can highly recommend it. Together with Marlin or Fluidd as UI it's much better than OctoPrint and Marlin imho. Don't go with OctoPrint if you wanna give Klipper a try tho (just my 2ct ;) ). Since you have a spare mobo now, you could make a test setup where you flash that mobo to get a first impression about how things look, feel and work.  

As for the fw-files: I just checked and at least the links at my infosite to where AC is offering them still work and you can find the 1.5.6.3 there.  
I also have the .4 because someone sent it to me - if you send me an email (address in the footer at each page of the infosite), I can send it to you if you want.   

Edit: 
I now added an expandable infobox about the GC6609 to the chapter "Mainboard" (right above the first picture). Could you please check if the image (from a Kobra Neo) fits your drivers? If not, could you maybe send me a picture I could use instead?

**Reply by kmars42** [reply]
HOST_ACTION_COMMANDS allow the printer to report its condition to OctoPrint such as whether you've paused or cancelled a print.  I've heard good things about Klipper and will probably go that way in the future but it's more than I have time to look into for now.

I looked at your image and it looks identical to mine except that the processor chip number appears to be GD32F103 while mine is GD32F303.  The resolution on yours isn't quite good enough for me to be sure though.  

I would love to have a copy of the newer firmware and will email you a picture of my motherboard without all of the cables being plugged into it and a close-up of the driver chip.  Thanks, and especially thank  you for making all of this information available!!

**Reply by GNN3** [reply]
> MEATPACK

In Configuration_adv.h:
#define MEATPACK_ON_SERIAL_PORT_1
//#define MEATPACK_ON_SERIAL_PORT_2

In Configuration.h:
#define SERIAL_PORT 2
Compression on an used port is blocked.

PS: I'm sorry, I didn't see the "_"

**Comment by GNN3** [comment]
We managed to assemble a workable firmware for the Kobra 2 NEO for the Arduino-MFL native framework, using the MCU at full speed (120 MHz), all available 96 KB of RAM and a hardware FPU.
This allows you to use the FT-motion motion handler (as in Klipper).
The code is still very raw, the main problem is that it is not yet possible to fully use DMA in the display driver, pixel-by-pixel rendering overloads the CPU.
Any help in bringing the code to an acceptable version is welcome.
https://github.com/GNN3/Marlin_K2N.git

**Reply by GNN3** [reply]
Fixes and optimizations of SPI for TFT: updated dividers, added synchronization barriers, improved data processing.
Fixed a bug with incorrect display modes ST7789.
At a speed of 300 mm/s with accelerations of the order of 5000, a decent print quality is obtained.
When starting the printer for the first time after flashing, it is necessary to park the printer head, and then start the automatic leveling.
The automatic Z-offset setting has not yet been implemented, it must be set manually via the menu.
PS: the best results are obtained with the aka Klipper profile in the slicer, and you also need to disable support for drawing arcs there (this only unnecessarily loads the processor).
The settings for the firmware retract and shapers are available in the menu.
